Sources/Task/TaskProgress.swift

```diff
@@ -29,87 +29,145 @@ private struct KVO {
+            for key in Observation.attributes {
+                observee.addObserver(self, forKeyPath: key.rawValue, options: Observation.options, context: &Observation.attributesContext)
+            }
+            observee.addObserver(self, forKeyPath: KeyPath.cancelled.rawValue, options: Observation.options, context: &Observation.cancelledContext)
+            observee.addObserver(self, forKeyPath: KeyPath.paused.rawValue, options: Observation.options, context: &Observation.pausedContext)
+
+            OSAtomicOr32Barrier(State.observing.rawValue, &state)
         }
 
-        if NSProgress.currentProgress()?.paused == true {
-            original.pause()
-        } else {
-            pausingHandler = original.pause
+        func cancelObserving(observee: NSProgress) {
+            let oldState = State(rawValue: UInt32(bitPattern: OSAtomicAnd32Orig(~State.ready.rawValue, &state)))
+            guard !oldState.isStrictSupersetOf(.cancellable) else { return }
+            OSAtomicOr32(State.cancelled.rawValue, &state)
+
+            for key in Observation.attributes {
+                observee.removeObserver(self, forKeyPath: key.rawValue, context: &Observation.attributesContext)
+            }
+            observee.removeObserver(self, forKeyPath: KeyPath.cancelled.rawValue, context: &Observation.cancelledContext)
+            observee.removeObserver(self, forKeyPath: KeyPath.paused.rawValue, context: &Observation.pausedContext)
         }
 
-        if #available(OSX 10.11, iOS 9.0, *) {
-            resumingHandler = original.resume
+        private override func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
+            guard let keyPath = keyPath where object != nil && (state & State.ready.rawValue) != 0, let observer = observer, newValue = change?[NSKeyValueChangeNewKey] else { return }
+            switch context {
+            case &Observation.cancelledContext:
+                observer.inheritCancelled(newValue as! Bool)
+            case &Observation.pausedContext:
+                observer.inheritPaused(newValue as! Bool)
+            case &Observation.attributesContext:
+                observer.inheritValue(newValue, forKeyPath: keyPath)
+            default:
+                preconditionFailure("Unexpected KVO context for private object")
+            }
         }
     }
 
-    func detach() {
-        for keyPath in KVO.KeyPath.all {
-            original.removeObserver(self, forKeyPath: keyPath.rawValue, context: &KVO.context)
+    let observee: NSProgress
+    var token: Observation?
+
+    init(attachingTo observee: NSProgress) {
+        let current = NSProgress.currentProgress()
+        self.observee = observee
+        super.init(parent: current, userInfo: nil)
+
+        if current?.cancelled == true {
+            observee.cancel()
+        }
+
+        if current?.paused == true {
```

Comment:
Do we expect it to be both canceled and paused simultaneously?

Is it legit to `pause` after we've already `cancel`ed?